File: hasPathSum.py
# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    accum = 0
    def firTravel(self,root,sum):
        self.accum += root.val
        print self.accum
        if self.accum == sum and root.left == None and root.right == None:
            return True
        if root.left:
            if self.firTravel(root.left,sum):
                return True
        if root.right:
            if self.firTravel(root.right,sum):
                return True
        self.accum -= root.val
        return False
    # @param root, a tree node
    # @param sum, an integer
    # @return a boolean
    def hasPathSum(self,root,sum):
        if root == None:
        	return False
        return self.firTravel(root,sum)
if __name__ == '__main__':
    Slu = Solution()
    a = TreeNode(2)
    b = TreeNode(3)
    c = TreeNode(4)
    a.left = b
    a.right = c
    print Slu.hasPathSum(a,6)


        

        

File: Cpp/hasPathSum.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

//一开始用了static变量，WA,leetcode每次都用同一个instance,static变量不会清零，慎用
class Solution {
public:
   bool hasPathSum(TreeNode * root , int sum){
	   if (root == NULL)
		   return false;
	   return firTravel(root,sum);
   }
   bool firTravel(TreeNode *root, int sum, int accum = 0) {
       if (root == NULL)
            return false;
       accum += root->val;
       if (accum == sum && !root->left && !root->right)
           return true;
       if (root->left)
           if (firTravel(root->left, sum,accum))
               return true;
       if (root->right)
           if (firTravel(root->right, sum,accum))
               return true;
       accum -= root->val;
       return false;
   }
};
